### PR TITLE
[Spec Decode] Enable adaptive DSpark on SM100 sparse MLA

### LINKED ISSUE #52785: [Feature][DSpark]: Enable varlen for GLM-5.x/DSV3.2 (Sparse MLA)
### 🚀 The feature, motivation and pitch

Adaptive verification has a firm requirement that the attention backend must not use the CPU-side query lengths as a source-of-truth, instead using the GPU tensors (so that we can run the confidence-based assignment on the GPU without synchronizing the CPU).

Additionally, kernels must be able to support decode batches where requests have different query lengths per request.

GLM-5.x models using DSV3.2-style sparse MLA should be compatible through FlashInfer's SM100 TRT-LLM sparse-MLA kernel. vLLM currently derives the sparse token-to-request mapping from CPU query boundaries, and the target capability check also includes draft attention that is captured by a separate graph manager. We should wire up the device-length path, validate correctness, evaluate performance, and merge support.

PROPOSED IN https://github.com/vllm-project/vllm/pull/52783

DEPENDS ON https://github.com/vllm-project/vllm/pull/47808

Parent tracker: https://github.com/vllm-project/vllm/issues/51303

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by mgoin**
## Purpose

Resolves #52785. This is part of the adaptive DSpark backend expansion tracked by #51303. Related GLM-5.2 DSpark bring-up context is in #50851.

This enables adaptive DSpark verification with `FlashInferMLASparseTRTLLMBackend`. It was validated end to end with `nvidia/GLM-5.2-NVFP4` and `RedHatAI/GLM-5.2-speculator.dspark` on 4× NVIDIA B300.

glm52_dspark_adaptive

Before this change, startup failed for two independent reasons:

1. The sparse-MLA metadata builder derived token-to-request mappings from CPU query boundaries. Adaptive verification changes those boundaries on device without synchronizing them back to the CPU.
2. The target graph capability check included the draft model's FlashAttention backend. Draft attention is captured and capability-checked by its own graph manager, and it never consumes the target's adaptive query lengths.

## Changes

- Add an SM100 FlashInfer sparse-MLA metadata builder that derives request IDs from device query boundaries and declares variable-length CUDA graph support.
- Scope target CUDA graph and query-length compatibility checks to target attention layers while retaining draft groups needed by draft prefill.
- Add a real B300 kernel regression using GLM-5.2 dimensions. It compares FlashInfer sparse MLA against SDPA with device query lengths `[1, 7, 3, 5]` and deliberately stale CPU lengths `[4, 4, 4, 4]`.
- Add a unit regression proving draft-only groups are excluded without hiding mixed target/draft groups that still participate in target execution.

## Performance

Scaled SPEED Bench qualitative sweep, 16K context, 512 output tokens, temperature 1.0, seed 0. The c1 run used 32 prompts, c16 used 128, and the remaining points used 256. Both configurations completed with zero failed requests.

| Concurrency | Fixed k=7 tok/s | Adaptive tok/s | Gain |
|---:|---:|---:|---:|
| 1 | 306.7 | 310.6 | +1.3% |
| 16 | 2,084.4 | 2,093.5 | +0.4% |
| 32 | 3,525.0 | 3,595.8 | +2.0% |
| 64 | 5,236.8 | 5,538.8 | +5.8% |
| 128 | 7,172.9 | 8,190.7 | +14.2% |
| 256 | 7,720.4 | 10,316.8 | +33.6% |

At c256, mean TPOT improved from 23.24 ms to 17.37 ms. A direct B300 kernel comparison also favored the selected FlashInfer SM100 backend over FlashMLA by 2.1–2.6× for uniform decode batches and 31.9× for a mixed 1/3/5/7-token adaptive shape.

## Model evaluation

GSM8K, five-shot, 200 examples, greedy decoding, 2,048-token output limit:

| Configuration | Accuracy | Invalid/truncated |
|---|---:|---:|
| Fixed k=7 | 89.0% | 1.0% |
| Adaptive k≤7 | 91.0% | 0.5% |

This shows no detected accuracy regression at this sample size; the two-point difference is not presented as statistically significant. HLE was also exercised, but 38/40 sampled outputs still truncated at an 8,192-token limit, so no HLE accuracy result is reported.

## Reproduction

```bash
chg run -g=4 -- vllm serve nvidia/GLM-5.2-NVFP4 \
  --tensor-parallel-size 4 \
  --enable-expert-parallel \
  --load-format instanttensor \
  --kv-cache-dtype fp8 \
  --block-size 64 \
  --max-model-len 16384 \
  --max-num-batched-tokens 16384 \
  --max-num-seqs 256 \
  --gpu-memory-utilization 0.8 \
  --compilation-config '{"max_cudagraph_capture_size":2048}' \
  --speculative-config '{"method":"dspark","model":"RedHatAI/GLM-5.2-speculator.dspark","attention_backend":"FLASH_ATTN","num_speculative_tokens":7,"draft_sample_method":"probabilistic","enable_adaptive_verification":true}'
```

The full server completed target and draft CUDA graph capture and served the benchmark/evaluation requests successfully.

## Test plan

```bash
chg run -g=1 -- .venv/bin/python -m pytest \
  tests/v1/attention/test_dspark_noncausal_sparse_mla.py -q
# 21 passed, 8 skipped

.venv/bin/python -m pytest tests/v1/worker/test_attn_utils.py -q
# 7 passed

.venv/bin/pre-commit run ruff-check --files <changed files>
.venv/bin/pre-commit run ruff-format --files <changed files>
.venv/bin/pre-commit run mypy-3.12 --hook-stage manual --files <runtime files>
# all passed
```

The focused CPU and B300 tests plus lint and mypy were rerun after rebasing onto current `main`.

## Not a duplicate

Duplicate checks covered #51303 and adaptive DSpark/sparse-MLA keywords. #52362 enables adaptive verification on ROCm and changes AMD model/indexer paths; it does not overlap this CUDA SM100 metadata and target/draft graph-scoping change. #50694 addresses earlier FA3 and DSpark/DFlash proposer wiring for GLM-5.2.

## AI assistance

AI assistance (OpenAI Codex) was used to investigate, implement, test, benchmark, and draft this change. The submitter reviewed every changed line, understands the change end to end, and ran the tests and model evaluations reported above.

### CODE REVIEW

File: vllm/v1/worker/gpu/model_runner.py [hunk 1]
```diff
@@ -531,8 +531,20 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
 
+        target_attn_layer_names = None
+        if isinstance(self.speculator, DraftModelSpeculator):
+            # Draft prefill reuses these groups, but draft attention is captured
+            # and capability-checked by its own graph manager.
+            target_attn_layer_names = {
+                layer_name
+                for group in self.kv_cache_config.kv_cache_groups
+                for layer_name in group.layer_names
+            } - self.speculator.draft_attn_layer_names
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu/model_runner.py:548]: this will disable CG check for the drafter completely (e.g. if there was a backend that was CG NEVER we would want to fall back to PIECEWISE); i think we should try to isolate this to `maybe_create_adaptive_verification_manager` if possible

File: tests/v1/attention/test_dspark_noncausal_sparse_mla.py [hunk 2]
```diff
@@ -290,6 +292,14 @@ def _sparse_sdpa(idx_tensor, q_tok, k_mqa=k_mqa, v_mqa=v_mqa):
     common_attn_metadata = create_common_attn_metadata(
         batch_spec, block_size, device, arange_block_indices=True
     )
+    if cpu_query_lens is not None:
+        assert len(cpu_query_lens) == len(query_lens)
+        cpu_query_start_loc = [0]
+        for query_len in cpu_query_lens:
+            cpu_query_start_loc.append(cpu_query_start_loc[-1] + query_len)
+        common_attn_metadata.query_start_loc_cpu = torch.tensor(
```
Comment (benchislett) [inline tests/v1/attention/test_dspark_noncausal_sparse_mla.py:302]: where is this consumed? I can't tell why this was added

**Review by benchislett — COMMENTED** [review]
LGTM, besides @LucasWilkinson's comment

File: vllm/v1/worker/gpu/model_runner.py (hunk 1 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/model_runner.py:548]: Good catch! I restored the global drafter check and made target check only for adaptive verification, with a test covering the fallback

File: tests/v1/attention/test_dspark_noncausal_sparse_mla.py (hunk 2 above)
Comment (mgoin) [reply tests/v1/attention/test_dspark_noncausal_sparse_mla.py:302]: This intentionally keeps the CPU boundaries stale because adaptive updates only the device boundaries. I renamed the variable to `stale_cpu_query_lens` and added a short comment clarifying

It gets consumed here https://github.com/vllm-project/vllm/blob/bd8865a299c4a68cff9b6443b9fd795f4c4735f6/vllm/v1/attention/backend.py#L597-L601

**Comment by gaby** [comment]
@mgoin Thank you for your all the improvements to GLM! 💪

**Review by LucasWilkinson — APPROVED** [review]
